### PR TITLE
Add explanation of how metadata is mapped

### PR DESCRIPTION
**Pull request by phynes-sensiblecode**
New markdown file that explains the relationship between the data
available in `cantabular-metadata` and the source CSV files.

### CODE REVIEW

File: CENSUS_METADATA.md [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+| `Geographic_Coverage` | `String!` | `Dataset.Geographic_Coverage` | `Dataset.Geographic_Coverage_Welsh` |
+| `Dataset_Population` | `String!` | `Dataset.Dataset_Population` | `Dataset.Population_Welsh` |
+| `Last_Updated` | `String` | `Dataset.Last_Updated` | |
+| `Unique_Url` | `String` | `Dataset.Unique_Url` | |
+| `Contact` | `Contact` | Keyed on `Dataset.Contact_Id` | |
+| `Version` | `String!` | `Dataset.Version` | |
+| `Related_Datasets` | `[String]!` | List of `Related_Dataset.Related_Dataset_Mnemonic` values keyed on `Related_Dataset.Dataset_Mnemonic` | |
+| `Keywords` | `[String]!` | List of `Dataset_Keyword.Dataset_Keyword` values keyed on `Dataset_Keyword.Dataset_Mnemonic` | List of `Dataset_Keyword.Dataset_Keyword_Welsh` values keyed on `Dataset_Keyword.Dataset_Mnemonic` |
+| `Publications` | `[Publication]!` | Keyed on `Publication_Dataset.Dataset_Mnemonic` | |
+| `Census_Releases` | `[Census_Release]!` | Keyed on `Release_Dataset.Dataset_Mnemonic`/`Census_Release_Number`| |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:143]:
```suggestion
| `Census_Releases` | `[Census_Release]!` | List of `Census_Release` values keyed on `Release_Dataset.Dataset_Mnemonic`/`Census_Release_Number`| |
```

File: CENSUS_METADATA.md [hunk 2]
```diff
@@ -0,0 +1,778 @@
...
+| `Statistical_Unit` | `Statistical_Unit!` | Keyed on `Dataset.Statistical_Unit` | |
+| `Alternate_Geographic_Variables` | `[String!]` | List of alternate geographic variable names which are available for this table sourced from `Dataset_Variables.csv` | |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:145]:
```suggestion
| `Alternate_Geographic_Variables` | `[String!]` | List of alternate geographic variable names which are available for this table sourced from `Dataset_Variable.csv` keyed on `Dataset_Variable.Dataset_Mnemonic` | |
```

File: CENSUS_METADATA.md [hunk 3]
```diff
@@ -0,0 +1,778 @@
...
+fields are sourced from `Database.csv`. The other fields in the file are contained in the `DatasetMetadata`
+object.
+
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `name` | `String!` | `Database.Database_Mnemonic` | |
+| `label` | `String` | `Database.Database_Title` | `Database.Database_Title_Welsh` |
+| `description` | `String` | `Database.Description` | `Database.Database_Description_Welsh` |
+| `meta` | `DatasetMetadata!` | Additional fields from `Dataset.csv` | |
+| `vars(names: [String!]!)` | `[Variable]!` | | |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:159]: Is the `(name: [String!]!)` bit intentional?

File: CENSUS_METADATA.md [hunk 4]
```diff
@@ -0,0 +1,778 @@
...
+The `Dataset` object is a built-in type in `cantabular-metadata`. The `name`, `label` and `description`
+fields are sourced from `Database.csv`. The other fields in the file are contained in the `DatasetMetadata`
+object.
+
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `name` | `String!` | `Database.Database_Mnemonic` | |
+| `label` | `String` | `Database.Database_Title` | `Database.Database_Title_Welsh` |
+| `description` | `String` | `Database.Description` | `Database.Database_Description_Welsh` |
+| `meta` | `DatasetMetadata!` | Additional fields from `Dataset.csv` | |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:158]: Typo, `Dataset.csv` should be `Database.csv`?

File: CENSUS_METADATA.md [hunk 5]
```diff
@@ -0,0 +1,778 @@
...
+| `Version` | `String!` | `Variable.Version` | |
+| `Statistical_Unit` | `Statistical_Unit` | Keyed on `Variable.Statistical_Unit` | |
+| `Keywords` | `[String]!` | List of `Variable_Keyword.Variable_Keyword` values keyed on `Variable_Keyword.Variable_Mnemonic` | List of `Variable_Keyword.Variable_Keyword_Welsh` values keyed on `Variable_Keyword.Variable_Mnemonic`|
+| `Topic` | `Topic` | Keyed on `Variable.Topic_Mnemonic` | |
+| `Questions` | `[Question]!` | Keyed on `Variable.Source_Question.Variable_Mnemonic`/`Source_Question_Code` | |
+| `Variable_Type` | `Variable_Type!` | `Variable.Variable_Type_Code` | |
+| `Quality_Statement_Text` | `String` | `Variable.Quality_Statement_Text` | |
+| `Quality_Summary_URL` | `String` | `Variable.Quality_Summary_URL` | |
+
+## Variable_Type
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:247]: Should be `###`?

File: CENSUS_METADATA.md [hunk 6]
```diff
@@ -0,0 +1,778 @@
...
+| `Comparability_Comments` | `String` | `Variable.Comparability_Comments` | `Variable.Comparability_Comments_Welsh` |
+| `Uk_Comparison_Comments` | `String` | `Variable.Uk_Comparison_Comments` | `Variable.Uk_Comparison_Comments_Welsh` |
+| `Geographic_Abbreviation` | `String` | `Variable.Geographic_Abbreviation` | `Variable.Geographic_Abbreviation_Welsh` |
+| `Geographic_Theme` | `String` | `Variable.Geographic_Theme` | `Variable.Geographic_Theme_Welsh` |
+| `Geographic_Coverage` | `String` | `Variable.Geographic_Coverage` | `Variable.Geographic_Coverage_Welsh` |
+| `Version` | `String!` | `Variable.Version` | |
+| `Statistical_Unit` | `Statistical_Unit` | Keyed on `Variable.Statistical_Unit` | |
+| `Keywords` | `[String]!` | List of `Variable_Keyword.Variable_Keyword` values keyed on `Variable_Keyword.Variable_Mnemonic` | List of `Variable_Keyword.Variable_Keyword_Welsh` values keyed on `Variable_Keyword.Variable_Mnemonic`|
+| `Topic` | `Topic` | Keyed on `Variable.Topic_Mnemonic` | |
+| `Questions` | `[Question]!` | Keyed on `Variable.Source_Question.Variable_Mnemonic`/`Source_Question_Code` | |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:242]:
```suggestion
| `Questions` | `[Question]!` | List of `Question` keyed on `Variable.Source_Question.Variable_Mnemonic`/`Source_Question_Code` | |
```

File: CENSUS_METADATA.md [hunk 7]
```diff
@@ -0,0 +1,778 @@
...
+
+The `Parent_Classification_Mnemonic`, `Signed_Off_Flag`, `Flat_Classification_Flag` and `Id` field in `Classification.csv` are ignored.
+
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `Mnemonic_2011` | `String` | `Classification.Mnemonic_2011` or `null` for geographic variables | |
+| `Default_Classification_Flag` | `String` | `Classification.Default_Classification_Flag` or `null` for geographic variables | |
+| `Version` | `String!` | `Classification.Version` or `Variable.Version` for geographic variables | |
+| `ONS_Variable` | `ONS_Variable!` | Keyed on `Classification.Variable_Mnemonic` or `Variable.Variable_Mnemonic` for geographic variables | |
+| `Topics` | `[Topic]!` | Keyed on `Topic_Classification.Classification_Mnemonic`/`Topic_Mnemonic` or `[]` for geographic variables | |
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:218]:
```suggestion
| `Topics` | `[Topic]!` | List of `Topic_Classification` keyed on `Topic_Classification.Classification_Mnemonic`/`Topic_Mnemonic` or `[]` for geographic variables | |
```

File: CENSUS_METADATA.md (hunk 5 above)
Comment (SpencerHedger) [reply CENSUS_METADATA.md:247]: I'm finding the heading sizes hard to follow, are they supposed to all be the same size or is there some purposeful size change happening that I'm missing?

File: CENSUS_METADATA.md [hunk 8]
```diff
@@ -0,0 +1,778 @@
...
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `Statistical_Unit` | `String!` | `Statistical_Unit.Statistical_Unit` | |
+| `Statistical_Unit_Description` | `String!` | `Statistical_Unit.Statistical_Unit_Description` | `Statistical_Unit.Statistical_Unit_Description_Welsh` |
+
+
+# Sample queries
+
+This section contains some examples of GraphQL queries that can be used to query metadata for the
+dataset name `Teaching-Dataset` from `cantabular-metadata` and `cantabular-api-ext`. Sample responses
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:345]:
```suggestion
dataset named `Teaching-Dataset` from `cantabular-metadata` and `cantabular-api-ext`. Sample responses
```

File: CENSUS_METADATA.md [hunk 9]
```diff
@@ -0,0 +1,778 @@
...
+
+```
+{
+  dataset(name: "Teaching-Dataset") {
+    name
+    description
+    meta {
+      Lowest_Geog_Variable
+    }
+    vars(names: ["Region" "Sex"]){
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:467]:
```suggestion
    vars(names: ["Region", "Sex"]){
```

File: CENSUS_METADATA.md [hunk 10]
```diff
@@ -0,0 +1,778 @@
...
+
+```
+{
+  dataset(name: "Teaching-Dataset" lang: "cy") {
+    name
+    description
+    meta {
+      Lowest_Geog_Variable
+    }
+    vars(names: ["Region" "Sex"]){
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:532]:
```suggestion
    vars(names: ["Region", "Sex"]){
```

File: CENSUS_METADATA.md [hunk 11]
```diff
@@ -0,0 +1,778 @@
...
+            "Age"
+          ]
+        }
+      ]
+    }
+  }
+}
+```
+
+### Get metadata for dataset and variables
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:452]:
```suggestion
### Get metadata for a dataset and its variables
```

File: CENSUS_METADATA.md [hunk 12]
```diff
@@ -0,0 +1,778 @@
...
+            ]
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+### Get metadata for dataset and variables in Welsh
```
Comment (SpencerHedger) [inline CENSUS_METADATA.md:518]:
```suggestion
### Get metadata for a dataset and its variables in Welsh
```

Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:518]: OK

File: CENSUS_METADATA.md (hunk 11 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:452]: OK

File: CENSUS_METADATA.md (hunk 9 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:467]: The syntax I have is valid. Commas are "Ignored Tokens":
https://spec.graphql.org/October2021/#sec-Language.Source-Text.Ignored-Tokens

However, I'll add for readability.

File: CENSUS_METADATA.md (hunk 10 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:532]: OK (see note above)

File: CENSUS_METADATA.md (hunk 5 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:247]: All those object types should be `##`. There appears to be a mix of `##` and `###`. Will change.

File: CENSUS_METADATA.md (hunk 4 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:158]: OK

File: CENSUS_METADATA.md (hunk 3 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:159]: I don't think so.

File: CENSUS_METADATA.md (hunk 2 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:145]: OK

File: CENSUS_METADATA.md [hunk 13]
```diff
@@ -0,0 +1,778 @@
...
+
+The schema of the metadata loaded into the service can be customised
+at runtime by supplying a GraphQL schema file along with the metadata
+content, allowing for a wide variety of metadata to be loaded.
+
+The software allows metadata to be attached to:
+ - a **service** (which serves a collection of datasets)
+ - the **datasets** within the service (from which tables can be built using queries)
+ - the **variables** which exist in each dataset
+ - pre-defined **tables**
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:22]: OK

File: CENSUS_METADATA.md [hunk 14]
```diff
@@ -0,0 +1,778 @@
...
+Metadata can be loaded in multiple languages by specifying a
+language parameter on input metadata JSON. This allows other software that
+makes use of the metadata service to be supplied with localised reference
+metadata, where relevant, for all metadata types.
+
+# Processing 2021 census metadata
+
+## CSV metadata files
+
+The CSV files contain metadata for:
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:33]: OK.

File: CENSUS_METADATA.md [hunk 15]
```diff
@@ -0,0 +1,778 @@
...
+
+Data from the CSV files are mapped to values in built-in `cantabular-metadata` objects and to
+user defined objects. The names of user-defined objects have been chosen to reflect the names of
+the source file names wherever possible. The field names within the user defined objects have also
+been chosen to reflect the column names in the source files.
+
+The source files contain some fields that are Welsh equivalents of other fields. The metadata service
+has separate objects for English and Welsh versions of the high level metadata objects.
+
+If an optional Welsh field is not populated, then the English version will typically be used in its place.
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:49]: Yes- the category labels as mentioned below.

File: CENSUS_METADATA.md [hunk 16]
```diff
@@ -0,0 +1,778 @@
...
+
+## Geography lookup file
+
+The main set of CSV files does not contain category information for geographic variables. A separate geography
+lookup file contains this information. These Python scripts extract the Welsh category labels for geographic
+variables from this file.
+
+# `cantabular-metadata` schema
+
+This section describes the `cantabular-metadata` schema and describes how fields in the output are
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:67]: OK

File: CENSUS_METADATA.md [hunk 17]
```diff
@@ -0,0 +1,778 @@
...
+derived from the source files. A table is provided for each object which details the field name,
+the GraphQL type and the sources for the English and Welsh versions of the metadata.
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:69]: I'll make "GraphQL" a hyperlink.

File: CENSUS_METADATA.md [hunk 18]
```diff
@@ -0,0 +1,778 @@
+# 2021 census metadata in Cantabular
+
+The scripts in this repository convert census metadata in CSV format to JSON files
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:3]: OK

File: CENSUS_METADATA.md [hunk 19]
```diff
@@ -0,0 +1,778 @@
...
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `meta` | `ServiceMetadata!` | | |
+| `tables` | `[Table!]` | Tables read from `Dataset.csv` | |
+
+## ServiceMetadata
+
+All the fields in `ServiceMetadata` are user defined. The census metadata consists of data that is
+relevant on a per dataset basis, so this object only has a `description` field containing a basic
+message.
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:91]: OK

File: CENSUS_METADATA.md [hunk 20]
```diff
@@ -0,0 +1,778 @@
...
+
+## TableMetadata
+
+All the fields in `TableMetadata` are user defined. It contains information from `Dataset.csv` that
+is not included in the `Table` object.
+
+The object also contains data from `Related_Dataset.csv`, `Dataset_Keyword.csv`, `Publication_Dataset.csv`
+and `Release_Dataset.csv`.
+
+The `Signed_Off_Flag` and `Id` field in `Classification.csv` are ignored.
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:129]: No. `Security_Mnemonic` is used to determine whether tables (and variables) are included in the Cantabular metadata. I'll add a note to this effect.

File: CENSUS_METADATA.md [hunk 21]
```diff
@@ -0,0 +1,778 @@
...
+| Field | GraphQL Type | Source (en) | Source (cy) |
+| --- | --- | --- | --- |
+| `Dataset_Mnemonic_2011` | `String` | `Dataset.Dataset_Mnemonic_2011` | |
+| `Geographic_Coverage` | `String!` | `Dataset.Geographic_Coverage` | `Dataset.Geographic_Coverage_Welsh` |
+| `Dataset_Population` | `String!` | `Dataset.Dataset_Population` | `Dataset.Population_Welsh` |
+| `Last_Updated` | `String` | `Dataset.Last_Updated` | |
+| `Unique_Url` | `String` | `Dataset.Unique_Url` | |
+| `Contact` | `Contact` | Keyed on `Dataset.Contact_Id` | |
+| `Version` | `String!` | `Dataset.Version` | |
+| `Related_Datasets` | `[String]!` | List of `Related_Dataset.Related_Dataset_Mnemonic` values keyed on `Related_Dataset.Dataset_Mnemonic` | |
```
[the remark this answers is not in the document]
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:140]: OK

File: CENSUS_METADATA.md (hunk 6 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:242]: OK

File: CENSUS_METADATA.md (hunk 7 above)
Comment (phynes-sensiblecode) [reply CENSUS_METADATA.md:218]: List of `Topic`...

**Review by SpencerHedger — APPROVED** [review]
Looks good to me.
